### PR TITLE
PutResult and DeleteResult allow 0 updated tables

### PR DESCRIPTION
**Pull request by nikitin-da**
Closes of #546 
@artem-zinnatullin PTAL, please

### CODE REVIEW

**Comment by artem-zinnatullin** [comment]
Please rebase on latest master

File: storio-sqlite/src/main/java/com/pushtorefresh/storio/sqlite/operations/put/PutResult.java [hunk 1]
```diff
@@ -127,7 +127,7 @@ public boolean wasNotInserted() {
      * even if all columns were same.
      */
     public boolean wasUpdated() {
-        return numberOfRowsUpdated != null;
+        return numberOfRowsUpdated != null && numberOfRowsUpdated > 0;
```
Comment (nikitin-da) [inline storio-sqlite/src/main/java/com/pushtorefresh/storio/sqlite/operations/put/PutResult.java]: It seems that we need it again (#423)
@artem-zinnatullin, I know you dislike it, but I think it is not true if we will return `wasUpdated() // true` if `0` rows were updated=)

Comment (artem-zinnatullin) [inline storio-sqlite/src/main/java/com/pushtorefresh/storio/sqlite/operations/put/PutResult.java]: Well, okay, but please fix javadoc too!

File: storio-sqlite/src/main/java/com/pushtorefresh/storio/sqlite/operations/delete/PreparedDeleteByQuery.java [hunk 2]
```diff
@@ -48,7 +48,9 @@
     public DeleteResult executeAsBlocking() {
         try {
             final DeleteResult deleteResult = deleteResolver.performDelete(storIOSQLite, deleteQuery);
-            storIOSQLite.internal().notifyAboutChanges(Changes.newInstance(deleteResult.affectedTables()));
+            if (deleteResult.numberOfRowsDeleted() > 0) {
```
Comment (artem-zinnatullin) [inline storio-sqlite/src/main/java/com/pushtorefresh/storio/sqlite/operations/delete/PreparedDeleteByQuery.java]: Thanks for fix!

File: storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/ObserveChangesDeleteTest.java [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+        assertThat(deleteResult.numberOfRowsDeleted()).isEqualTo(1);
+
+        emissionChecker.assertThatNoExpectedValuesLeft();
+
+        subscription.unsubscribe();
+    }
+
+    @Test
+    public void shouldNotNotifyIfWasNotDeletedObject() {
+        final EmissionChecker emissionChecker = new EmissionChecker(new LinkedList<Changes>());
```
Comment (artem-zinnatullin) [inline storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/ObserveChangesDeleteTest.java]: If you change `new LinkedList` to `emptyList()` from `Collections` test will be even more obvious and easy to understand!

File: storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/ObserveChangesDeleteTest.java [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+        }
+
+        emissionChecker.assertThatNoExpectedValuesLeft();
+
+        subscription.unsubscribe();
+    }
+
+    @Test
+    public void shouldNotNotifyIfWasNotDeletedCollectionWithTransaction() {
+        final EmissionChecker emissionChecker = new EmissionChecker(new LinkedList<Changes>());
```
Comment (artem-zinnatullin) [inline storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/ObserveChangesDeleteTest.java]: Same about `emptyList()`

File: storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/ObserveChangesDeleteTest.java [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+
+        // Should not receive changes
+        emissionChecker.assertThatNoExpectedValuesLeft();
+
+        subscription.unsubscribe();
+    }
+
+    @Test
+    public void shouldNotNotifyIfWasNotDeletedCollectionWithoutTransaction() {
+        final EmissionChecker emissionChecker = new EmissionChecker(new LinkedList<Changes>());
```
Comment (artem-zinnatullin) [inline storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/ObserveChangesDeleteTest.java]: `emptyList()`

File: storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/ObserveChangesDeleteTest.java (hunk 3 above)
Comment (artem-zinnatullin) [inline storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/ObserveChangesDeleteTest.java]: Oh, got it, constructor requires `Queue`, just ignore my comments about `emtyList()` then :)

File: storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/ObserveChangesDeleteTest.java [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+import java.util.List;
+import java.util.Queue;
+
+import rx.Subscription;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@RunWith(RobolectricGradleTestRunner.class)
+@Config(constants = BuildConfig.class, sdk = 21)
+public class ObserveChangesDeleteTest extends BaseTest {
```
Comment (artem-zinnatullin) [inline storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/ObserveChangesDeleteTest.java]: Awesome set of tests!

File: storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/ObserveChangesPutTest.java [hunk 7]
```diff
@@ -0,0 +1,530 @@
...
+
+        subscription.unsubscribe();
+    }
+
+    @Test
+    public void shouldNotifyAboutInsertCollectionWithoutTransaction() {
+        final List<User> users = TestFactory.newUsers(10);
+
+        final Queue<Changes> expectedChanges = new LinkedList<Changes>();
+        for (int i = 0; i < users.size(); ++i) {
```
Comment (artem-zinnatullin) [inline storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/ObserveChangesPutTest.java]: `++i` for the speed? :smile_cat:

File: storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/ObserveChangesPutTest.java [hunk 8]
```diff
@@ -0,0 +1,530 @@
...
+        // Update
+        emissionChecker.awaitNextExpectedValue();
+
+        emissionChecker.assertThatNoExpectedValuesLeft();
+
+        subscription.unsubscribe();
+    }
+
+    @Test
+    public void shouldNotifyAboutUpdateCollectionWithOutTransaction() {
```
Comment (artem-zinnatullin) [inline storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/ObserveChangesPutTest.java]: `WithOut` -> `Without`

File: storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/UserTableMeta.java [hunk 9]
```diff
@@ -79,7 +82,160 @@ public PutResult performPut(@NonNull StorIOSQLite storIOSQLite, @NonNull User ob
...
+                    final ContentValues contentValues = mapToContentValues(object);
+
+                    if (cursor.getCount() == 0) {
+                        final InsertQuery insertQuery = mapToInsertQuery(object);
+                        final long insertedId = storIOSQLite.internal().insert(insertQuery, contentValues);
+                        putResult = PutResult.newInsertResult(insertedId, insertQuery.table());
+                    } else {
+                        boolean existObjectsEqual = true;
+                        cursor.moveToFirst();
+                        while (cursor.moveToNext()) {
```
Comment (artem-zinnatullin) [inline storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/UserTableMeta.java]: Looks like you're skipping first value, or I'm wrong?

File: storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/UserTableMeta.java [hunk 10]
```diff
@@ -79,7 +82,160 @@ public PutResult performPut(@NonNull StorIOSQLite storIOSQLite, @NonNull User ob
...
```
Comment (artem-zinnatullin) [inline storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/UserTableMeta.java]: Same question about first value

File: storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/UserTableMeta.java [hunk 11]
```diff
@@ -105,4 +261,8 @@ public DeleteQuery mapToDeleteQuery(@NonNull User user) {
     private UserTableMeta() {
         throw new IllegalStateException("No instances please");
     }
+
+    public static ContentValues toContentValues(User user) {
```
Comment (artem-zinnatullin) [inline storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/UserTableMeta.java]: Nullability

File: storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/ObserveChangesPutTest.java (hunk 7 above)
Comment (nikitin-da) [inline storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/ObserveChangesPutTest.java]: Exactly! We just add a few slow integration tests, so as a compensation..

File: storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/UserTableMeta.java (hunk 9 above)
Comment (nikitin-da) [inline storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/UserTableMeta.java]: Oh, you are right

**Comment by nikitin-da** [comment]
Fixed!

**Comment by artem-zinnatullin** [comment]
LGTM! Please fix the javadoc and we'll merge it :)

**Comment by nikitin-da** [comment]
Integration -> Unit =) It will be faaast!

**Comment by artem-zinnatullin** [comment]
PR now much smaller!

**Comment by nikitin-da** [comment]
It seems, for content resolver we just can to fix only `put` and `delete` results

**Comment by artem-zinnatullin** [comment]
okay

File: storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/operations/delete/DeleteStub.java [hunk 12]
```diff
@@ -162,14 +210,15 @@ private void verifyTransactionBehavior() {
             verify(internal).endTransaction();
 
             // Only one notification should be thrown
```
Comment (artem-zinnatullin) [inline storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/operations/delete/DeleteStub.java]: This comment is now invalid

File: storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/operations/delete/PreparedDeleteByQueryTest.java [hunk 13]
```diff
@@ -158,4 +158,37 @@ public void shouldWrapExceptionIntoStorIOExceptionObservable() {
...
+        final DeleteQuery deleteQuery = DeleteQuery.builder()
+                .table("test_table")
+                .where("column1 = ?")
+                .whereArgs(1)
+                .build();
+
+        //noinspection unchecked
+        final DeleteResolver<DeleteQuery> deleteResolver = mock(DeleteResolver.class);
+
+        final DeleteResult expectedDeleteResult = DeleteResult.newInstance(0, deleteQuery.table()); // No items was deleted
```
Comment (artem-zinnatullin) [inline storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/operations/delete/PreparedDeleteByQueryTest.java]: `were` deleted

**Comment by nikitin-da** [comment]
Fixed comments

**Comment by artem-zinnatullin** [comment]
:+1:

**Pull request merged**